### ISSUE TITLE
Duplicate PRIMARY during concurrency

### ISSUE THREAD

**Issue by jakechan2012** [dotnetcore/CAP#294]
```text
fail: DotNetCore.CAP.DefaultSubscriberExecutor[0]
      An exception occurred while executing the subscription method. Topic:events.activity_create, Id:1103205991968251904
DotNetCore.CAP.Internal.SubscriberExecutionFailedException: Duplicate entry '1103205992291213312' for key 'PRIMARY' ---> MySql.Data.MySqlClient.MySqlException: Duplicate entry '1103205992291213312' for key 'PRIMARY' ---> MySql.Data.MySqlClient.MySqlException: Duplicate entry '1103205992291213312' for key 'PRIMARY'
   at MySqlConnector.Core.ServerSession.TryAsyncContinuation(Task`1 task) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ServerSession.cs:line 1248
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at MySqlConnector.Core.ResultSet.ReadResultSetHeaderAsync(IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\Core\ResultSet.cs:line 42
   --- End of inner exception stack trace ---
   at MySql.Data.MySqlClient.MySqlDataReader.ActivateResultSet(ResultSet resultSet) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 80
   at MySql.Data.MySqlClient.MySqlDataReader.ReadFirstResultSetAsync(IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 302
   at MySql.Data.MySqlClient.MySqlDataReader.CreateAsync(MySqlCommand command, CommandBehavior behavior, ResultSetProtocol resultSetProtocol, IOBehavior ioBehavior) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlDataReader.cs:line 287
   at MySqlConnector.Core.TextCommandExecutor.ExecuteReaderAsync(String commandText, MySqlParameterCollection parameterCollection, CommandBehavior behavior, IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\Core\TextCommandExecutor.cs:line 37
   at MySql.Data.MySqlClient.MySqlCommand.ExecuteNonQueryAsync(IOBehavior ioBehavior, CancellationToken cancellationToken) in C:\projects\mysqlconnector\src\MySqlConnector\MySql.Data.MySqlClient\MySqlCommand.cs:line 261
   at Dapper.SqlMapper.ExecuteImplAsync(IDbConnection cnn, CommandDefinition command, Object param) in C:\projects\dapper\Dapper\SqlMapper.Async.cs:line 678
   at DotNetCore.CAP.MySql.MySqlPublisher.ExecuteAsync(CapPublishedMessage message, ICapTransaction transaction, CancellationToken cancel)
   at DotNetCore.CAP.Abstractions.CapPublisherBase.PublishAsyncInternal(CapPublishedMessage message)
   at DotNetCore.CAP.MySql.MySqlPublisher.PublishCallbackAsync(CapPublishedMessage message)
   at DotNetCore.CAP.Internal.CallbackMessageSender.SendAsync(String messageId, String topicName, Object bodyObj)
   at DotNetCore.CAP.DefaultSubscriberExecutor.InvokeConsumerMethodAsync(CapReceivedMessage receivedMessage)
   --- End of inner exception stack trace ---
   at DotNetCore.CAP.DefaultSubscriberExecutor.InvokeConsumerMethodAsync(CapReceivedMessage receivedMessage)
   at DotNetCore.CAP.DefaultSubscriberExecutor.ExecuteWithoutRetryAsync(CapReceivedMessage message)
```

**Comment by jakechan2012** [comment]
这个情况应该是我的应用负载均衡，但共用一个数据库引起的，但是 Subscriber 无法多线程订阅，事件排队执行肯定无法满足我的性能需求，我只能通过启动多个进程来解决，配置不同数据库还是加表前缀对我的部署都是一个挑战 ☹️

**Comment by yang-xiaodong** [comment]
Please provide your configuration and environment

**Comment by jakechan2012** [comment]
@yang-xiaodong 我现在通过环境变量为每个应用分配了表前缀，并发下没有再出现主键问题

配置如下：
```csharp
cap.UseMySql(mysql =>
{
    mysql.ConnectionString = options.MySqlConnectionString;
    mysql.TableNamePrefix = options.MySqlTableNamePrefix;
});
cap.UseRabbitMQ(mq =>
{
    mq.ExchangeName = options.RabbitMQExchangeName;
    mq.HostName = options.RabbitMQHostName;
    mq.UserName = options.RabbitMQUserName;
    mq.Password = options.RabbitMQPassword;
});
cap.DefaultGroup = options.Group;
cap.Version = options.Version;
cap.FailedRetryCount = options.FailedRetryCount;
cap.FailedRetryInterval = options.FailedRetryInterval;
cap.FailedThresholdCallback = options.FailedCallback;
cap.UseDashboard();
```

服务器是 Ubuntu Server 16.04, 8GB, i5

**Comment by yang-xiaodong** [comment]
which version are u using？

**Comment by yang-xiaodong** [comment]
We use the snowflake algorithm to generate the primary key, you can find the code [here]( https://github.com/dotnetcore/CAP/blob/master/src/DotNetCore.CAP/Infrastructure/SnowflakeId.cs#L57), we randomize the workerId and datacenterId , the probability of theoretical conflict is very low, but this possibility is not ruled out

**Comment by jakechan2012** [comment]
@yang-xiaodong the latest version: 2.4.2

**Comment by yang-xiaodong** [comment]
Hello @jakechan2012 , we will fixed the issus at v2.5.0

You can set up the snowflake algorithm generation by adding `CAP_WORKERID` and `CAP_DATACENTERID` to the environment variables to avoid concurrency conflicts.

https://github.com/dotnetcore/CAP/blob/8773ec6ea4b2dfc32a5d7d38ff7139d6c36a5bcf/src/DotNetCore.CAP/Infrastructure/SnowflakeId.cs#L59-L67

**Comment by jakechan2012** [comment]
@yang-xiaodong many thx, but I prefer the option `TableNamePrefix ` (both need set up env variables)

**Comment by yang-xiaodong** [comment]
Fixed in v2.5.0
